jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
+    Function will check string against Process.name(), Process.exe() and Process.cmdline()
 
-    Raises:
-        j.exceptions.RuntimeError: [description]
-        j.exceptions.NotImplemented: [description]
+    Args:
+        process_name (str): The target process name
+        match_predicate (callable, optional): Function that does matching between found processes and the targeted process.
+            the function should accept two arguments and return a boolean. Defaults to None.
+        limit (int, optional): If not equal to 0, function will return as fast as the number of PID(s) found become equal to `limit` value.
+        _alt_source(callable or iterable, optional): Can be used to specify an alternative source of the psutil.Process objects to match against.
+            ex: get_user_processes func, or get_similar_processes.
+            if not specified, psutil.process_iter will be used. Defaults to None.
+        include_zombie (bool, optional): Whether to include pid for zombie proccesses or not. Defaults to False.
+        full_cmd_line (bool, optional): The pattern is normally only matched against the process name.
+            if full_cmd_line is set to True, the full command line is used. Defaults to False.
 
     Returns:
-        [list(int)] -- list of pids
+        list of int: list of PID(s)
     """
     # default match predicate
-    # why aren't we using psutil ??
     def default_predicate(target, given):
-        return target.strip().lower() in given.lower()
+        if isinstance(given, list):
+            return target in given
+        else:
+            return target.strip().lower() == given.lower()
 
-    if match_predicate is None:
-        match_predicate = default_predicate
+    default_processes_source = psutil.process_iter(["name", "exe", "cmdline"])
 
-    if process_name is None:
-        raise j.exceptions.RuntimeError("process cannot be None")
-    if j.data.platform.is_unix():
-        pids = set()
-        for process in get_processes():
-            try:
-                pid = process.pid
-                if not isinstance(pid, int):
-                    continue
-                name = process.name()
-                if match_predicate(process_name, name):
-                    pids.add(pid)
-                elif match_predicate(process_name, process.exe()):
-                    pids.add(pid)
-                else:
-                    cmdline = process.cmdline()
-                    if cmdline and cmdline[0]:
-                        if match_predicate(process_name, cmdline[0]):
-                            pids.add(pid)
-            except (psutil.Error, FileNotFoundError):
+    match_predicate = match_predicate or default_predicate
+    p_source = _alt_source or default_processes_source
+
+    pids = []
+    for proc in p_source:
+        try:
+            if not include_zombie and proc.status() == psutil.STATUS_ZOMBIE:
+                j.logger.debug(f"ignoring : {proc.pid} zombie process.")
```

Comment:
IMHO, these inner debug logs like this can be removed after making sure everything works.